language-support/ts/daml-ledger/index.test.ts

```diff
@@ -0,0 +1,244 @@
+    await server.connected;
+    console.log("connected");
+    stream.on("change", (evs) => console.log(evs));
+    stream.on("close", (ev) => console.log(ev));
+    stream.off("change", (evs) => console.log(evs));
+    stream.off("close", (ev) => console.log(ev));
+    server.send({ errors: ["not good!"] });
+  });
+
+  test("receive empty events", async () => {
+    let receivedEvents: object[] = [];
+    const server = new WS("ws://localhost:4000/v1/stream/query", {
+      jsonProtocol: true,
+    });
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    await server.connected;
+    console.log("connected");
+    stream.on("change", (evs) => (receivedEvents = [...evs]));
+    stream.on("close", (ev) => console.log(ev));
+    stream.off("change", (evs) => console.log(evs));
+    stream.off("close", (ev) => console.log(ev));
+    server.send({ events: [] });
+    expect(receivedEvents).toEqual([]);
+  });
+
+  test("receive one event", async () => {
+    let receivedEvents: object[] = [];
+    const server = new WS("ws://localhost:4000/v1/stream/query", {
+      jsonProtocol: true,
+    });
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    await server.connected;
+    console.log("connected");
+    stream.on("change", (evs) => (receivedEvents = [...evs]));
+    stream.on("close", (ev) => console.log(ev));
+    stream.off("change", (evs) => console.log(evs));
+    stream.off("close", (ev) => console.log(ev));
+    server.send({ events: [fooEvent(1)] });
+    expect(receivedEvents).toEqual([fooCreateEvent(1)]);
+  });
+
+  test("receive several events", async () => {
+    let receivedEvents: object[] = [];
+    const server = new WS("ws://localhost:4000/v1/stream/query", {
+      jsonProtocol: true,
+    });
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    await server.connected;
+    console.log("connected");
+    stream.on("change", (evs) => (receivedEvents = [...evs]));
+    stream.on("close", (ev) => console.log(ev));
+    stream.off("change", (evs) => console.log(evs));
+    stream.off("close", (ev) => console.log(ev));
+    server.send({ events: [1, 2, 3].map(fooEvent) });
+    expect(receivedEvents).toEqual([1, 2, 3].map(fooCreateEvent));
+  });
+
+  test("drop matching created and archived events", async () => {
+    let receivedEvents: object[] = [];
+    const server = new WS("ws://localhost:4000/v1/stream/query", {
+      jsonProtocol: true,
+    });
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    await server.connected;
+    console.log("connected");
+    stream.on("change", (evs) => (receivedEvents = [...evs]));
+    stream.on("close", (ev) => console.log(ev));
+    stream.off("change", (evs) => console.log(evs));
+    stream.off("close", (ev) => console.log(ev));
+    server.send({ events: [fooEvent(1), fooEvent(2), fooArchiveEvent(1)] });
```

Comment:
The JSON API will never send us `create` and `archive` events for the same contract id within one `events` message. We should spread this over two separate messages.

Reply:
the server now sends create/archive events in two separate messages.